Use multiple registry fallbacks instead of one

wilzbach opened the pull request:
As we already see sometimes errors on Travis while downloading the installer script due to dlang.org and nightlies.dlang.org being down, it's imho a good idea to have multiple fallbacks by default for DUB.

https://code-mirror2.dlang.io is maintained by myself and the Heroku-App is based on https://github.com/dlang/dub-registry/pull/231

wilzbach commented:
Btw we didn't we simply add a list of RegistryPackageSuppliers as default?
A lot of the methods of `Dub` look like this:

```d
foreach(ps; m_packageSuppliers){
	try {
		auto versions = ps.getVersions(p);
```

Doing the fallbacking as "Meta" supplier as done now and improved by this PR _and_ within the class itself seems redundant.

MartinNowak commented:
> Btw we didn't we simply add a list of RegistryPackageSuppliers as default?

This was discussed in the original PR, https://github.com/dlang/dub/pull/1190#issuecomment-320616565.

s-ludwig commented:
Is it possible to set up a CNAME record for codemirror3.dlang.io that points to the Heroku app, or do they do any internal dispatching based on the original name? It would be good to keep a central place where the mirrors can be replaced/removed, so that we don't have to rely on application updates until such modifications take effect.

wilzbach commented:
> Is it possible to set up a CNAME record for codemirror3.dlang.io that points to the Heroku app, or do they do any internal dispatching based on the original name?

Custom CNames are free at Heroku, but SSL for them isn't. We could, however, setup a wildcard redirect. dlang.io can be managed via Namecheap - I will invite you later today.

s-ludwig commented:
A wildcard redirect would work. Regarding switching to HTTPS for code.dlang.org. As a simple first step we could have `["https://code.dlang.org/", "http://code.dlang.org/", ...]` to provide the fallback in case any client has problems with HTTPS.

wilzbach commented:
> but uses the same DNS as code-mirror.dlang.io (another failure point, e.g. misconfiguration)

@s-ludwig I added the Heroku App as code-mirror3.dlang.io, but I have to agree with @MartinNowak: by using the Heroku DNS, we have another fallback in case NameCheap has some issues (also wildcard redirect don't seem to be possible directly with NameCheap)

But I did add "http://code.dlang.org/" as a fallback if SSL isn't supported and fixed the Travis errors for 2.066 (as GDC is at 2.074.0 in master and 2.068.0 with their last release, we could drop the support for 2.066)

s-ludwig commented:
Alternatively, we could add some other mechanism for maintaining the mirror list in the future, such as requesting an up-to-date one from code.dlang.org, which would ideally be signed to avoid a worthwhile security target.  BTW, package signing is also something we should also put on the (virtual) higher priority list as the eco system keeps growing.

MartinNowak commented:
Let's not sweat this too much, adding a few mirrors is nice as it might help with certain issues.
But this mirror thing isn't a replacement for a proper HA setup (or PaaS) in the long-term.

MartinNowak commented on source/dub/packagesupplier.d:
```diff
@@ -292,17 +292,19 @@ class RegistryPackageSupplier : PackageSupplier {
 	{
-		return format("%s (fallback %s)", m_default.description, m_fallback.description);
+		import std.algorithm : map;
+		return format("%s (fallback %s)", m_default.description, m_fallbacks.map!(x => x.description));
```
Not a blocker, but this is fairly long and detailed.
```
==== registry at https://code.dlang.org/ (fallback ["registry at http://code.dlang.org/", "registry at https://code-mirror.dlang.io/", "registry at https://code-mirror2.dlang.io/", "registry at https://dub-registry.herokuapp.com/"]) ====
```

wilzbach replied:
I know, but the `PackageSupplier` interface currently only supports a `description` attribute. We could add sth. like `location` in a follow-up?

MartinNowak commented on source/dub/packagesupplier.d:
```diff
@@ -322,7 +324,18 @@ package alias FallbackPackageSupplier = AutoImplement!(AbstractFallbackPackageSu
+				try
+					return m_fallback.%1$s(args);
+				catch(Throwable)
+				    assert(1);
```
Isn't `{}` what you should use for an empty statement?

s-ludwig replied:
More importantly, `Throwable` is a bad idea, must be `Exception` - logging the error with `logDiagnostic` or maybe just `logDebug` could also be important for remote debugging .

MartinNowak commented on source/dub/packagesupplier.d:
```diff
@@ -322,7 +324,18 @@ package alias FallbackPackageSupplier = AutoImplement!(AbstractFallbackPackageSu
+				catch(Throwable)
+				    assert(1);
+			}
+			return m_fallbacks.back.%1$s(args);
```
The recursive structure was setup to chain exceptions, so that users get a list of unreachable sites. Not too important though.

MartinNowak commented on source/dub/dub.d:
```diff
@@ -80,7 +86,7 @@ PackageSupplier[] defaultPackageSuppliers()
 		new FallbackPackageSupplier(
 			new RegistryPackageSupplier(URL(defaultRegistryURL)),
-			new RegistryPackageSupplier(URL(fallbackRegistryURL))
+			fallbackRegistryURLs.map!(x => cast(PackageSupplier) new RegistryPackageSupplier(URL(x))).array
```
Could have stayed with the recursive structure.
```d
fallbackRegistryURLs.reduce!((agg, url) => new FallbackPackageSupplier(agg, new RegistryPackageSupplier(URL(x))))
```

wilzbach replied:
Hmm, I thought the "linear" way is more intuitive, but I don't really mind. If you feel strongly the recursive structure, feel free to change it back.

wilzbach commented:
> Just nits, lets move on with this.

Addressed the nits.

> The recursive structure was setup to chain exceptions, so that users get a list of unreachable sites. Not too important though.

Don't they get this with `logDebug` as well?
Anyhow as mentioned before if you prefer the recursive way, feel free to change.
In any case it would be great to get this in before the next release.
